### PR TITLE
Hassh feature 2698 v4

### PR DESCRIPTION
**Pull request by MalakhatkoVadym**
Link to redmine ticket:
https://redmine.openinfosecfoundation.org/issues/2698

Describe changes:
- Adds hassh and hasshServer for ssh fingerprinting to suricata.
- Adds lua logging to get hassh fingerprints.

Modifies #5003 by 
1. Removed ssh prefix eve-json-format.rst;
2. Added proper git author;
3. Fixed git messages;
4. Replaced unsafe calls to C code with the atomic bool variable.
5. Made rust code more clear.
6. Moved tests to suricata-verify. https://github.com/OISF/suricata-verify/pull/247
7. Moved id's of hassh errors to the end of the list instead of inserting them

### Testing

1. Enable hassh in suricata.yaml in app-layer.protocols.ssh.hassh:

```
app-layer:
  protocols:
    ssh:
      enabled: yes
      hassh: yes
```

2. Check your hassh parameters (hassh, server hassh, hassh string, hassh server string) (connect with ssh to tested host) using tool https://github.com/salesforce/hassh/tree/master/python 
or using suricata's lua scripting:
change suricata.yaml:
```
outputs:
  - lua:
      enabled: yes
      scripts-dir: /etc/suricata/lua-output/
      scripts:
         - hassh.lua
```
hassh.lua:

```
function init (args)
    local needs = {}
    needs["protocol"] = "ssh"
    return needs
end

function setup (args)
    filename = "/tmp/hassh.log"
    file = assert(io.open(filename, "a"))
end

function log(args)
    hassh = HasshGet()
    file:write (hassh .. " -hassh \n")

    hasshServer = HasshServerGet()
    file:write (hasshServer .. " -hasshServer \n")

    hasshString = HasshGetString()
    file:write (hasshString .. " -hasshString \n")

    hasshServerString = HasshServerGetString()
    file:write (hasshServerString .. " -hasshServerString \n")

    file:flush()
end

function deinit (args)
    file:close(file)
end

return 0
```

3. Add rules for hassh detection:
```
alert ssh any any -> any any (msg:"match SSH hash"; ssh.hassh; content:"xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"; sid:1000010;)
alert ssh any any -> any any (msg:"match SSH hash-server"; ssh.hassh.server; content:"xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"; sid:1000020;)
alert ssh any any -> any any (msg:"match SSH hash-string"; ssh.hassh.string; content:"none,zlib@openssh.com,zlib"; sid:1000030;)
alert ssh any any -> any any (msg:"match SSH hash-server-string"; ssh.hassh.server.string; content:"umac-64-etm@openssh.com,umac-128-etm@openssh.com,"; sid:1000040;)
```
4. Check alerts.

### CODE REVIEW

File: doc/userguide/output/eve/eve-json-format.rst [hunk 1]
```diff
@@ -989,10 +995,14 @@ Example of SSH logging:
     "client": {
         "proto_version": "2.0",
         "software_version": "OpenSSH_6.7",
+        "ssh.hassh": "ec7378c1a92f5a8dde7e8b7a1ddf33d1",
+        "ssh.hassh.string": "curve25519-sha256,diffie-hellman-group14-sha256,diffie-hellman-group14-sha1,ext-info-c",
      },
     "server": {
         "proto_version": "2.0",
         "software_version": "OpenSSH_6.7",
+        "ssh.hassh.server": "ec7378c1a92f5a8dde7e8b7a1ddf33d1",
```
Comment (catenacyber) [inline doc/userguide/output/eve/eve-json-format.rst:1004]: Maybe I do not remember the discussion.
But it still feels strange/redundant to have `ssh.hassh.server` in `server` block

Can we have only `ssh.hassh` and `ssh.hassh.string` and then use the direction ?

File: rust/Cargo.toml.in [hunk 2]
```diff
@@ -28,6 +28,7 @@ num = "0.2"
 num-derive = "0.2"
 num-traits = "0.2"
 widestring = "0.4"
+md5 = "0.7.0"
```
Comment (catenacyber) [inline rust/Cargo.toml.in:31]: @jasonish opinions on this ?

performance discussed here https://github.com/OISF/suricata/pull/4974#discussion_r432895304

If we choose to go with this, should we create a ticket to get rid of libnss and move to Rust crypto crates ?

File: rust/src/ssh/logger.rs [hunk 3]
```diff
@@ -35,6 +35,18 @@ fn log_ssh(tx: &SSHTransaction) -> Option<Json> {
                 &tx.cli_hdr.swver,
             );
         }
+        if tx.cli_hdr.hassh.len() > 0 {
```
Comment (catenacyber) [inline rust/src/ssh/logger.rs:38]: Can you rebase this with jsonbuilder ? (mostly adding `?`)

File: rust/src/ssh/parser.rs [hunk 4]
```diff
@@ -92,10 +128,89 @@ named!(pub ssh_parse_record<SshRecordHeader>,
...
+        };
+        // reserving memory
+        hassh_string.reserve_exact(slices.iter().fold(0, |acc, x| acc + x.len()));
+        // copying slices to hassh string
+        slices.iter().for_each(|&x| hassh_string.extend_from_slice(x)); 
+        hassh.extend(format!("{:x?}", compute(&hassh_string)).as_bytes());
+    }
+}
+
+named!(parse_string<&[u8]>, do_parse!(
```
Comment (catenacyber) [inline rust/src/ssh/parser.rs:178]: For reference : https://securitylab.github.com/research/network-integers-are-hot-lava

File: rust/src/ssh/ssh.rs [hunk 5]
```diff
@@ -179,7 +194,7 @@ impl SSHState {
                             let remlen = rem.len() as u32;
                             hdr.record_left = head.pkt_len - 2 - remlen;
                             //header with rem as incomplete data
-                            if head.msg_code == SSH_MSG_NEWKEYS {
+                            if head.msg_code == parser::MessageCode::SshMsgNewKeys {
```
Comment (catenacyber) [inline rust/src/ssh/ssh.rs:197]: Did you try to split the TCP packet with the hassh data in 2 packets ?

I think that this case should be handled here somehow

File: rust/src/ssh/ssh.rs [hunk 6]
```diff
@@ -525,7 +540,7 @@ pub unsafe extern "C" fn rs_ssh_register_parser() {
         get_tx_detect_flags: Some(rs_ssh_get_tx_detect_flags),
         set_tx_detect_flags: Some(rs_ssh_set_tx_detect_flags),
     };
-
+
```
Comment (catenacyber) [inline rust/src/ssh/ssh.rs:543]: unneeded space

File: src/app-layer-ssh.c [hunk 7]
```diff
@@ -53,6 +53,24 @@
...
+/* HASSH fingerprints are disabled by default */
+#define SSH_CONFIG_DEFAULT_HASSH 0
+
+/**
+ * \brief if not explicitly disabled in config, enable hassh support
+ *
+ * Implemented using atomic to allow rule reloads to do this at
+ * runtime.
+ */
+void SSHEnableHassh(void)
```
Comment (catenacyber) [inline src/app-layer-ssh.c:65]: Maybe we do not need this wrapper function

File: rust/src/applayer.rs [hunk 8]
```diff
@@ -239,6 +239,11 @@ extern {
     pub fn AppLayerProtoDetectConfProtoDetectionEnabled(ipproto: *const c_char, proto: *const c_char) -> c_int;
 }
 
+// Defined in app-layer-ssh.h
+extern {
+    pub fn SSHHasshIsEnabled() -> bool;
```
Comment (catenacyber) [inline rust/src/applayer.rs:244]: Do we need this when it only wraps `rs_ssh_hassh_is_enabled` ?

File: src/output-lua.c [hunk 9]
```diff
@@ -823,8 +825,8 @@ static OutputInitResult OutputLuaLogInit(ConfNode *conf)
         } else if (opts.alproto == ALPROTO_SSH) {
             om->TxLogFunc = LuaTxLogger;
             om->alproto = ALPROTO_SSH;
-            om->tc_log_progress = SshStateBannerDone;
-            om->ts_log_progress = SshStateBannerDone;
+            om->tc_log_progress = SshStateFinished;
```
Comment (catenacyber) [inline src/output-lua.c:828]: Why this change ?

File: rust/src/ssh/logger.rs (hunk 3 above)
Comment (catenacyber) [reply rust/src/ssh/logger.rs:38]: Once #5050 is merged ;-)

File: rust/src/ssh/ssh.rs (hunk 5 above)
Comment (catenacyber) [reply rust/src/ssh/ssh.rs:197]: Yes, it looks like this pcap is good for testing.

File: src/output-lua.c (hunk 9 above)
Comment (catenacyber) [reply src/output-lua.c:828]: That is a good idea.
But then, there remains the question wether we lose the ability to use LUA for `ssh.proto_version` if we do not reach `SshStateKexDone` ? (for instance, just banner grabbing)

Comment (MalakhatkoVadym) [reply src/output-lua.c:828]: My bad, if to have code like this:
```suggestion
            om->tc_log_progress = SshStateBannerDone;
            om->ts_log_progress = SshStateFinished;;
```
`ssh.proto_version` will be printed using lua for the corrupted ssh session. And hassh parameters will be printed for full version.

Comment (MalakhatkoVadym) [reply src/output-lua.c:828]: @catenacyber @victorjulien I can add for now `if` for hassh and documentation that we can lose the ability to use LUA for ssh.proto_version if we do not reach `SshStateFinished`.
```suggestion
            if(rs_ssh_hassh_is_enabled())
                om->tc_log_progress = SshStateFinished;
            else
                om->tc_log_progress = SshStateBannerDone;
```

And I can add ticket for future to have smart `tc_log_progress` and `ts_log_progress` detection (callback function, which checks last state or ability to have few states at the same time)

Comment (MalakhatkoVadym) [reply src/output-lua.c:828]: @catenacyber @victorjulien 
Also, if I have code like this:
```suggestion
            om->tc_log_progress = SshStateFinished;
            om->ts_log_progress = SshStateFinished;;
```
It works for incomplete (I tested session with only banner packets) ssh session and `ssh.proto.version` is shown.

Comment (victorjulien) [reply src/output-lua.c:828]: I wonder if its worth exploring use of `OutputModule::TxLogCondition` here. Than you can have a bit more advanced logic.

Comment (catenacyber) [reply src/output-lua.c:828]: Thanks for the pointer

File: rust/src/ssh/parser.rs (hunk 4 above)
Comment (MalakhatkoVadym) [reply rust/src/ssh/parser.rs:178]: I tested for the incorrect length of fields and incorrect strings in kex message, parsing worked fine. Added unit test.

**Comment by MalakhatkoVadym** [comment]
https://github.com/OISF/suricata/pull/5067

**Comment by MalakhatkoVadym** [comment]
https://github.com/OISF/suricata/pull/5067

File: rust/src/ssh/parser.rs (hunk 4 above)
Comment (catenacyber) [reply rust/src/ssh/parser.rs:178]: Your code seems fine to me indeed with respect to this :-)
But then, the first referenced patch seemed fine to me (and to the author of the post)
So, thanks for double checking
